File: OJ practices/LeetCode题目集/39. Combination Sum(medium)-按序进入以去重.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.LinkedList;
import java.util.List;

/*
 * @Author: SourDumplings
 * @Date: 2020-02-13 22:17:46
 * @Link: https://github.com/SourDumplings/
 * @Email: changzheng300@foxmail.com
 * @Description: https://leetcode-cn.com/problems/combination-sum/
 */

class Solution
{
    private LinkedList<Integer> thisList = new LinkedList<>();
    private int thisSum = 0;
    private List<List<Integer>> res = new ArrayList<>();
    private int l;

    public List<List<Integer>> combinationSum(int[] candidates, int target)
    {
        l = candidates.length;
        Arrays.sort(candidates);
        for (int i = 0; i < l; i++)
        {
            backtrack(candidates, target, i);
        }
        return res;
    }

    private void backtrack(int[] candidates, int target, int start)
    {
        int thisNum = candidates[start];
        thisSum += thisNum;
        thisList.add(thisNum);
        if (thisSum == target)
        {
            res.add((LinkedList<Integer>) (thisList.clone()));
        }
        else if (thisSum < target)
        {
            for (int i = start; i < l; i++)
            {
                backtrack(candidates, target, i);
            }
        }
        thisSum -= thisNum;
        thisList.removeLast();
    }
}

File: OJ practices/LeetCode题目集/39. Combination Sum(medium)-按序进入以去重.py
'''
@Author: SourDumplings
@Date: 2020-02-14 16:30:53
@Link: https://github.com/SourDumplings/
@Email: changzheng300@foxmail.com
@Description: https://leetcode-cn.com/problems/combination-sum/
'''
from typing import List


class Solution:
    __res = []
    __thisList = []
    __l = 0

    def combinationSum(self, candidates: List[int], target: int) -> List[
            List[int]]:
        self.__l = len(candidates)
        candidates.sort()
        for i in range(self.__l):
            thisNum = candidates[i]
            if target < thisNum:
                break
            self.__thisList.append(thisNum)
            self.__backtrack(candidates, target - thisNum, i)
            self.__thisList.pop()
        return self.__res

    def __backtrack(self, candidates: List[int], target: int, start: int):
        if target == 0:
            self.__res.append(self.__thisList.copy())
        elif target > 0:
            for i in range(start, self.__l):
                thisNum = candidates[i]
                if target < thisNum:
                    break
                self.__thisList.append(thisNum)
                self.__backtrack(candidates, target - thisNum, i)
                self.__thisList.pop()
